fs: littlefs: Block device cache size improvements

DNedic opened the pull request:
Instead of picking a set of safe block multiple sized values for caches, round down to the nearest block multiple, increasing static buffer usage determined by Kconfig variables.

de-nordic reviewed (changes requested):
Change the commit title, you have previously used the "fs: littlefs:" prefix, so use that one or the "fs/littlefs:", do not introduce a new one.

DNedic commented on subsys/fs/littlefs_fs.c:
```diff
@@ -844,8 +844,19 @@ static int littlefs_init_cfg(struct fs_littlefs *fs, int flags)
-		lfs_size_t new_cache_size = block_size;
-		lfs_size_t new_lookahead_size = block_size * 4;
+		if (cache_size < block_size) {
+			LOG_ERR("Configured cache size is smaller than block size: %d < %d, "
+				"it must be a multiple of block size", cache_size, block_size);
```
Thanks! I will reduce the comment and remove the moves.

DNedic commented on subsys/fs/littlefs_fs.c:
```diff
@@ -844,8 +844,19 @@ static int littlefs_init_cfg(struct fs_littlefs *fs, int flags)
+		}
+		lcp->cache_size = ROUND_DOWN(cache_size, block_size);
+
+		if (lookahead_size < block_size) {
```
True, this could be an unexpected change. I thought 4 * blocksize was chosen as an arbitrary "safe" value. What is the underlying reason for it needing to be at least 4 * block size?

de-nordic replied:
That is a good question! The problem is that since we do not have answer to why * 4, since it was introduced actually,  I was assuming that you have some analysis done and now you have the answer.
Since we both are not able to answer the question, we should probably leave this as is and put some time into trying to figure out what is happening before changing that.

DNedic replied:
Yup, for this reason I have left the check as is for a minimal size, but the used size will be bigger if it can.

de-nordic replied:
I still agree that the hidden multiply, without explanation, should not happen here. If this is required, for some reason that we have not established, this should be imposed on user to set properly.
Specifically that the lookahead is supposed to store bitmap which, according to explanation here https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/fs/Kconfig.littlefs#L53, is supposed to compact info on 8 blocks into one byte, so lookahead is not really tied to block_size, rather to device size. But there is also comment that lookahead should be multiple of 8.

de-nordic replied:
OK, so, check me on that, but I do not see any logic behind the * 4 except for some tuning for user that has provided the patch in the first place.
I have looked here: https://github.com/zephyrproject-rtos/littlefs/blob/009bcff0ed4853a53df8256039fa815bda6854dd/lfs.c#L633 (and several other places) and it seems that the * 4 is not needed for anything. It also seems that the lookahead buffer  to hold info on no more than all blocks of area (https://github.com/zephyrproject-rtos/littlefs/blob/009bcff0ed4853a53df8256039fa815bda6854dd/lfs.c#L4350), which would be area_size / block_size / 8.

The size provided here is in bytes, each byte holding info on eight blocks, so also the Kconfig info on the need for the value of lookahead buffer to be multiple of 8 (https://github.com/zephyrproject-rtos/zephyr/blob/b7b3449f482cd9fb736e3a79363d390fd648051d/subsys/fs/Kconfig.littlefs#L59) is wrong.

DNedic replied:
Unless I am wrong here, there is no lower bound for lookahead size and the upper bound is block size / 8 anyway.

de-nordic replied:
So the Help for Kconfig is wrong and the * 4 is not needed either. So, should we then fix both here?

DNedic replied:
This should be fixed now, PTAL.

DNedic committed:
fs: littlefs: Improve block device cache size

Instead of picking a set of safe block multiple sized values for caches,
round down to the nearest block multiple, increasing static buffer usage
determined by Kconfig variables.

Signed-off-by: Djordje Nedic <nedic.djordje2@gmail.com>

DNedic committed:
fs: littlefs: Fix lookahead buffer sizing

The lookahead buffer is a bitmap containing info on free blocks and does
not have a minimum size or an alignment requirement.

This fixes the arbitrary 4 * block_size requirement for block devices
and fixes Kconfig documentation to remove a reference to nonexistent
8 byte alignment requirement.

Signed-off-by: Djordje Nedic <nedic.djordje2@gmail.com>

de-nordic reviewed (approved):
Looks OK. Thanks!
